### PR TITLE
Create new write repeating byte method

### LINKED ISSUE #1495: Add support for writing repeated bytes into a byte buffer.
I was working on a unit test where I needed `2**24` bytes in a byte buffer. Writing this the naive way is: `buffer.writeBytes(repeatElement(0, count: 1 << 24))`. This took around 16s. I replaced this with an unsafe pointer call to `memset` and it took more like 8ms.

We could solve this with a `writeRepeatedByte(_:count:)` that hides the unsafe code.

### PR DESCRIPTION
**Pull request by Davidde94**
Resolves #1495 

Add a function to support writing an arbitrary number of identical bytes.

### Motivation:

Sometimes it's necessary to do this, see #1495 

### Modifications:

Add a function that allocates `n` bytes, and uses `memset` to write a provided `UInt8` value to each byte. Also added a test.

### Result:

It's possible to quickly write `n` identical bytes to a buffer.
`buffer.writeRepeatingByte(9, count: 1 << 100)` (if you have enough memory)

### CODE REVIEW

File: Sources/NIO/ByteBuffer-aux.swift [hunk 1]
```diff
@@ -373,6 +373,22 @@ extension ByteBuffer {
...
+    /// Writes `byte` `repeated` times. Moves the writer index forward by the number of bytes written.
+    ///
+    /// - parameter byte: The `UInt8` byte to repeat.
+    /// - parameter count: How many times to repeat the given `byte`
+    /// - returns: How many bytes were written.
+    @discardableResult
+    @inlinable
+    public mutating func writeRepeatingByte(_ byte: UInt8, count: Int) -> Int {
+        let pointer = malloc(count)
+        memset(pointer!, Int32(byte), count)
```
Comment (weissi) [inline Sources/NIO/ByteBuffer-aux.swift:386]: You’re allocating memory here, then setting then bytes, and then copying them all into the target buffer. We can skip the first and the third operation (allocation & copy) and write straight into the target.

Check out [`withUnsafeWritableBytes(minimumWritableBytes:)`](https://apple.github.io/swift-nio/docs/current/NIO/Structs/ByteBuffer.html#/s:3NIO10ByteBufferV27writeWithUnsafeMutableBytes015minimumWritableH0_S2i_SiSwKXEtKF)
M

Comment (weissi) [reply Sources/NIO/ByteBuffer-aux.swift:386]: Actually, [`writeWithUnsafeWritableBytes(minimumWritableBytes:)`](https://apple.github.io/swift-nio/docs/current/NIO/Structs/ByteBuffer.html#/s:3NIO10ByteBufferV27writeWithUnsafeMutableBytes015minimumWritableH0_S2i_SiSwKXEtKF) is even better, that moves the writer index for you too.

File: Sources/NIO/ByteBuffer-aux.swift [hunk 2]
```diff
@@ -373,6 +373,22 @@ extension ByteBuffer {
...
     }
+    
+    /// Writes `byte` `repeated` times. Moves the writer index forward by the number of bytes written.
+    ///
+    /// - parameter byte: The `UInt8` byte to repeat.
+    /// - parameter count: How many times to repeat the given `byte`
+    /// - returns: How many bytes were written.
+    @discardableResult
+    @inlinable
+    public mutating func writeRepeatingByte(_ byte: UInt8, count: Int) -> Int {
```
Comment (weissi) [inline Sources/NIO/ByteBuffer-aux.swift:384]: We also need a precondition that this is `>= 0`

File: Sources/NIO/ByteBuffer-aux.swift (hunk 1 above)
Comment (Davidde94) [reply Sources/NIO/ByteBuffer-aux.swift:386]: Thanks a lot - I was looking for a function like that but clearly missed it

Comment (Davidde94) [reply Sources/NIO/ByteBuffer-aux.swift:386]: As for the leak - oops, just idiocy on my part

File: Sources/NIO/ByteBuffer-aux.swift [hunk 3]
```diff
@@ -373,6 +373,21 @@ extension ByteBuffer {
...
         return written
     }
+    
+    /// Writes `byte` `repeated` times. Moves the writer index forward by the number of bytes written.
+    ///
+    /// - parameter byte: The `UInt8` byte to repeat.
+    /// - parameter count: How many times to repeat the given `byte`
+    /// - returns: How many bytes were written.
+    @discardableResult
+    @inlinable
```
Comment (weissi) [inline Sources/NIO/ByteBuffer-aux.swift:383]: can we take off this `@inlinable`? The parameters here are `UInt8` and `Int`, if anything, inlining this would make it worse.

File: Sources/NIO/ByteBuffer-aux.swift [hunk 4]
```diff
@@ -373,6 +373,21 @@ extension ByteBuffer {
...
+    public mutating func writeRepeatingByte(_ byte: UInt8, count: Int) -> Int {
+        precondition(count >= 0, "Can't write less than 0 bytes")
+        return self.writeWithUnsafeMutableBytes(minimumWritableBytes: count) { (_pointer) -> Int in
```
Comment (weissi) [inline Sources/NIO/ByteBuffer-aux.swift:386]: - extra set of parens
- `_pointer` is underscored, should be `ptr` or `pointer`

File: Tests/NIOTests/ByteBufferTest.swift [hunk 5]
```diff
@@ -2579,6 +2579,21 @@ class ByteBufferTest: XCTestCase {
         self.buf.moveWriterIndex(forwardBy: 5)
         XCTAssertEqual("hellohello", self.buf.readString(length: 10))
     }
+    
+    func testRepeatingBytes() {
+
```
Comment (weissi) [inline Tests/NIOTests/ByteBufferTest.swift:2584]: we could make this test more readable by writing:

```
func testRepeatingBytes() {
    let written = self.buf.writeRepeatingByte(9, count: 1024)
    XCTAssertEqual(1024, written)
    XCTAssertEqual(Array(repeating: UInt8(9), count: 1024), Array(self.buf.readableBytesView))
}
```

- we should also add a test case that tests the edge condition (`count: 0`)

File: Sources/NIO/ByteBuffer-aux.swift [hunk 6]
```diff
@@ -373,6 +373,21 @@ extension ByteBuffer {
...
     }
+    
+    /// Writes `byte` `repeated` times. Moves the writer index forward by the number of bytes written.
+    ///
+    /// - parameter byte: The `UInt8` byte to repeat.
+    /// - parameter count: How many times to repeat the given `byte`
+    /// - returns: How many bytes were written.
+    @discardableResult
+    @inlinable
+    public mutating func writeRepeatingByte(_ byte: UInt8, count: Int) -> Int {
```
Comment (weissi) [inline Sources/NIO/ByteBuffer-aux.swift:384]: usually when we have a `write`, we also have a corresponding `set` method

File: Tests/NIOTests/ByteBufferTest.swift (hunk 5 above)
Comment (Davidde94) [reply Tests/NIOTests/ByteBufferTest.swift:2584]: Looks good, I'll also bump the number of bytes up I think

File: Sources/NIO/ByteBuffer-aux.swift (hunk 4 above)
Comment (Davidde94) [reply Sources/NIO/ByteBuffer-aux.swift:386]: Extra set of parens?

File: Sources/NIO/ByteBuffer-aux.swift [hunk 7]
```diff
@@ -373,6 +373,34 @@ extension ByteBuffer {
...
     }
+    
+    /// Writes `byte` `repeated` times. Moves the writer index forward by the number of bytes written.
+    ///
+    /// - parameter byte: The `UInt8` byte to repeat.
+    /// - parameter count: How many times to repeat the given `byte`
+    /// - returns: How many bytes were written.
+    @discardableResult
+    public mutating func writeRepeatingByte(_ byte: UInt8, count: Int) -> Int {
+        precondition(count >= 0, "Can't write less than 0 bytes")
```
Comment (Lukasa) [inline Sources/NIO/ByteBuffer-aux.swift:384]: I mean, it couldn’t possibly matter less, but...

```suggestion
        precondition(count >= 0, "Can't write fewer than 0 bytes")
```

File: Sources/NIO/ByteBuffer-aux.swift [hunk 8]
```diff
@@ -373,6 +373,34 @@ extension ByteBuffer {
...
+    }
+    
+    /// Sets the given `byte` `repeated` times at the given `index`. Will reserve more memory if necessary. Does not move the writer index.
+    ///
+    /// - parameter byte: The `UInt8` byte to repeat.
+    /// - parameter count: How many times to repeat the given `byte`
+    /// - returns: How many bytes were written.
+    @discardableResult
+    public mutating func setRepeatingByte(_ byte: UInt8, count: Int, at index: Int) -> Int {
+        precondition(count >= 0, "Can't write less than 0 bytes")
```
Comment (Lukasa) [inline Sources/NIO/ByteBuffer-aux.swift:398]: Again, not that it matters in the slightest.

```suggestion
        precondition(count >= 0, "Can't write fewer than 0 bytes")
```

Comment (Lukasa) [inline Sources/NIO/ByteBuffer-aux.swift:398]: This method is missing a `isKnownUniquelyReferenced` check, so it doesn’t correctly maintain the value semantics of the type. Take a look at `_setBytes` for an example of what needs to be done here.

Comment (Lukasa) [reply Sources/NIO/ByteBuffer-aux.swift:398]: The fact that this is passing its tests also means that there are no tests to confirm that this function actually does maintain the value semantics of the type: mind adding some?

File: Sources/NIO/ByteBuffer-aux.swift [hunk 9]
```diff
@@ -373,6 +373,34 @@ extension ByteBuffer {
         self._moveWriterIndex(forwardBy: written)
         return written
     }
+    
+    /// Writes `byte` `repeated` times. Moves the writer index forward by the number of bytes written.
```
Comment (glbrntt) [inline Sources/NIO/ByteBuffer-aux.swift:377]:
```suggestion
    /// Writes `byte` `count` times. Moves the writer index forward by the number of bytes written.
```

File: Sources/NIO/ByteBuffer-aux.swift [hunk 10]
```diff
@@ -373,6 +373,34 @@ extension ByteBuffer {
...
+    ///
+    /// - parameter byte: The `UInt8` byte to repeat.
+    /// - parameter count: How many times to repeat the given `byte`
+    /// - returns: How many bytes were written.
+    @discardableResult
+    public mutating func writeRepeatingByte(_ byte: UInt8, count: Int) -> Int {
+        precondition(count >= 0, "Can't write less than 0 bytes")
+        return self.writeWithUnsafeMutableBytes(minimumWritableBytes: count) { (pointer) -> Int in
```
Comment (glbrntt) [inline Sources/NIO/ByteBuffer-aux.swift:385]: Can this just be written in terms of `setRepeatingByte` to remove some duplication?

File: Tests/NIOTests/ByteBufferTest.swift [hunk 11]
```diff
@@ -2579,6 +2579,37 @@ class ByteBufferTest: XCTestCase {
...
+            let written = self.buf.writeRepeatingByte(9, count: count)
+            XCTAssertEqual(count, written)
+            XCTAssertEqual(Array(repeating: UInt8(9), count: count), Array(self.buf.readableBytesView))
+        }
+        
+        write(count: 1_000_000)
+        write(count: 0)
+    }
+    
+    func testSetBytes() {
```
Comment (glbrntt) [inline Tests/NIOTests/ByteBufferTest.swift:2595]: nit: this should including "repeating" (`setBytes` is its own thing)

File: Sources/NIO/ByteBuffer-aux.swift [hunk 12]
```diff
@@ -373,6 +373,34 @@ extension ByteBuffer {
...
+            memset(pointer.baseAddress!, Int32(byte), count)
+            return count
+        }
+    }
+    
+    /// Sets the given `byte` `repeated` times at the given `index`. Will reserve more memory if necessary. Does not move the writer index.
```
Comment (glbrntt) [inline Sources/NIO/ByteBuffer-aux.swift:391]:
```suggestion
    /// Sets the given `byte` `count` times at the given `index`. Will reserve more memory if necessary. Does not move the writer index.
```

File: Sources/NIO/ByteBuffer-aux.swift (hunk 8 above)
Comment (Lukasa) [reply Sources/NIO/ByteBuffer-aux.swift]: This thread is not outdated: it still has not been fixed.

Comment (Davidde94) [reply Sources/NIO/ByteBuffer-aux.swift]: GH marked it as outdated automatically because there was another change.

Comment (Davidde94) [reply Sources/NIO/ByteBuffer-aux.swift]: (When it's done I'll mark as resolved)

File: Sources/NIO/ByteBuffer-aux.swift [hunk 13]
```diff
@@ -373,6 +373,39 @@ extension ByteBuffer {
...
+    ///
+    /// - parameter byte: The `UInt8` byte to repeat.
+    /// - parameter count: How many times to repeat the given `byte`
+    /// - returns: How many bytes were written.
+    @discardableResult
+    public mutating func setRepeatingByte(_ byte: UInt8, count: Int, at index: Int) -> Int {
+        precondition(count >= 0, "Can't write fewer than 0 bytes")
+        let bytesCount = count
+        let newEndIndex: _Index = UInt32(index) + _toIndex(bytesCount)
+        if !isKnownUniquelyReferenced(&self._storage) {
```
Comment (weissi) [inline Sources/NIO/ByteBuffer-aux.swift:400]: - We shouldn't do the CoW-thing manually here yet, this is too dangerous.
- Please use

```swift
self.reserveCapacity(index + count)
self.withVeryUnsafeMutableBytes { destCompleteStorage in
            assert(destCompleteStorage.count >= index + count)
            let dest = destCompleteStorage[index ..< index + count]
            memset(dest.baseAddress!, byte, count: count)
        }
        return count
}
```

File: Sources/NIO/ByteBuffer-aux.swift [hunk 14]
```diff
@@ -373,6 +373,37 @@ extension ByteBuffer {
...
     }
+    
+    /// Writes `byte` `count` times. Moves the writer index forward by the number of bytes written.
+    ///
+    /// - parameter byte: The `UInt8` byte to repeat.
+    /// - parameter count: How many times to repeat the given `byte`
+    /// - returns: How many bytes were written.
+    @discardableResult
+    public mutating func writeRepeatingByte(_ byte: UInt8, count: Int) -> Int {
+        precondition(count >= 0, "Can't write fewer than 0 bytes")
```
Comment (weissi) [inline Sources/NIO/ByteBuffer-aux.swift:384]: kill this precondition please, it's repeated in `set` just below

File: Sources/NIO/ByteBuffer-aux.swift [hunk 15]
```diff
@@ -373,6 +373,37 @@ extension ByteBuffer {
...
+    /// - returns: How many bytes were written.
+    @discardableResult
+    public mutating func setRepeatingByte(_ byte: UInt8, count: Int, at index: Int) -> Int {
+        precondition(count >= 0, "Can't write fewer than 0 bytes")
+        self.reserveCapacity(index + count)
+        self.withVeryUnsafeMutableBytes { pointer in
+            precondition(pointer.count >= count)
+            let dest = pointer[index ..< index + count].base
+            memset(dest.baseAddress!, Int32(byte), count)
+//            memset(<#T##__b: UnsafeMutableRawPointer!##UnsafeMutableRawPointer!#>, <#T##__c: Int32##Int32#>, <#T##__len: Int##Int#>)
```
Comment (weissi) [inline Sources/NIO/ByteBuffer-aux.swift:403]: leftover comment

File: Sources/NIO/ByteBuffer-aux.swift [hunk 16]
```diff
@@ -373,6 +373,35 @@ extension ByteBuffer {
...
+    ///
+    /// - parameter byte: The `UInt8` byte to repeat.
+    /// - parameter count: How many times to repeat the given `byte`
+    /// - returns: How many bytes were written.
+    @discardableResult
+    public mutating func setRepeatingByte(_ byte: UInt8, count: Int, at index: Int) -> Int {
+        precondition(count >= 0, "Can't write fewer than 0 bytes")
+        self.reserveCapacity(index + count)
+        self.withVeryUnsafeMutableBytes { pointer in
+            precondition(pointer.count >= count)
```
Comment (weissi) [inline Sources/NIO/ByteBuffer-aux.swift:399]: this `precondition` is in the wrong place. It needs to be right in front of `memset` asserting the correct pointer has enough bytes left.

Comment (weissi) [reply Sources/NIO/ByteBuffer-aux.swift:399]: Probably something like

```swift
let dest = pointer[index...].base
preconditon(dest.count >= count)
memset(dest.baseAddress!, CInt(byte), size_t(count))
```

@Davidde94 also make sure to use the `CInt` (and not `Int32`) constructor as well as the `size_t` one. If you don't you're assuming the following things to be true:

- `CInt` == `Int32`
- `size_t` == `Int`

those are true for 64bit macOS but not on 32 bit systems usually.

**Comment by weissi** [comment]
hit #1445 on 5.0

**Comment by weissi** [comment]
@swift-nio-bot test this please

File: Sources/NIO/ByteBuffer-aux.swift [hunk 17]
```diff
@@ -373,6 +373,34 @@ extension ByteBuffer {
...
+    /// - parameter byte: The `UInt8` byte to repeat.
+    /// - parameter count: How many times to repeat the given `byte`
+    /// - returns: How many bytes were written.
+    @discardableResult
+    public mutating func setRepeatingByte(_ byte: UInt8, count: Int, at index: Int) -> Int {
+        precondition(count >= 0, "Can't write fewer than 0 bytes")
+        self.reserveCapacity(index + count)
+        self.withVeryUnsafeMutableBytes { pointer in
+            let dest = pointer.baseAddress?.advanced(by: index)
+            _ = dest!.initializeMemory(as: UInt8.self, repeating: byte, count: count)
```
Comment (weissi) [inline Sources/NIO/ByteBuffer-aux.swift:400]: can you move the `!` up to `pointer.baseAddress!`? Otherwise we may be doing extra work.

Comment (Davidde94) [reply Sources/NIO/ByteBuffer-aux.swift:400]: Done

File: Tests/NIOTests/DatagramChannelTests.swift [hunk 18]
```diff
@@ -249,10 +249,7 @@ final class DatagramChannelTests: XCTestCase {
             // will cause EMSGSIZE.
             let bufferSize = 1024 * 5
             var buffer = self.firstChannel.allocator.buffer(capacity: bufferSize)
-            buffer.writeWithUnsafeMutableBytes(minimumWritableBytes: bufferSize) {
-                _ = memset($0.baseAddress!, 4, $0.count)
-                return $0.count
-            }
+            buffer.writeRepeatingByte(4, count: buffer.capacity)
```
Comment (weissi) [inline Tests/NIOTests/DatagramChannelTests.swift:252]: this isn't a correct translation of the old code. Should be `count: bufferSize`. `ByteBufferAllocator` may over-allocate.

File: Tests/NIOTests/ByteBufferTest.swift [hunk 19]
```diff
@@ -723,21 +723,16 @@ class ByteBufferTest: XCTestCase {
         XCTAssertEqual(otherBuf, buf)
 
         // Write to both buffers.
-        let firstResult = buf!.withUnsafeMutableWritableBytes { (ptr: UnsafeMutableRawBufferPointer) -> Bool in
-            XCTAssertEqual(cap, ptr.count)
-            memset(ptr.baseAddress!, 0, ptr.count)
-            return false
-        }
-        let secondResult = otherBuf!.withUnsafeMutableWritableBytes { (ptr: UnsafeMutableRawBufferPointer) -> Bool in
+        self.buf.setRepeatingByte(0, count: self.buf.capacity, at: 0)
```
Comment (weissi) [inline Tests/NIOTests/ByteBufferTest.swift:726]: - `at: 0` isn't right, that should be `buf.writerIndex`
- `count: self.buf.capacity` also isn't right, that should be `buf.capacity - buf.writerIndex`

File: Tests/NIOTests/ByteBufferTest.swift [hunk 20]
```diff
@@ -723,21 +723,16 @@ class ByteBufferTest: XCTestCase {
         XCTAssertEqual(otherBuf, buf)
 
         // Write to both buffers.
-        let firstResult = buf!.withUnsafeMutableWritableBytes { (ptr: UnsafeMutableRawBufferPointer) -> Bool in
```
Comment (weissi) [inline Tests/NIOTests/ByteBufferTest.swift:726]: actually, sorry, this needs to be undone. This test is specifically testing `withUnsafeMutableWritableBytes`

**Comment by Davidde94** [comment]
@glbrntt ping
